### PR TITLE
Table cell formatting with a Cell class

### PR DESCRIPTION
**Pull request by takluyver**
This is an idea for how the context file mechanism could expose more flexible presentation of its output. Variable functions can return a `Cell` object:

```python
from damnit_ctx import Variable, Cell
import numpy as np

@Variable(summary='sum')
def var2(run):
    return Cell(np.arange(10), bold=False)
```

Returning `Cell(x)` with no extra parameters should behave the same as returning `x` directly, but extra parameters can control formatting (this is not yet implemented in the frontend). E.g. you could colour the background according to the value. You can also provide a custom summary value to show in the cell - i.e. you're no longer limited to single Numpy functions to reduce data.

Future parameters could also influence how we present the un-summarised data when double-clicking on the cell; although that might be better set in `@Variable()`.

### CODE REVIEW

**Comment by takluyver** [comment]
image

**Review by CammilleCC — COMMENTED** [review]
On first impression, it looks good! Thanks for the addition, would be happy to conveniently fetch the attributes (possible from the API?)

Would you already have ideas on how this extends on stored data display?

File: damnit/ctxsupport/damnit_ctx.py [hunk 1]
```diff
@@ -69,3 +79,67 @@ def annotations(self):
...
+        d = {}
+        if self.summary is not None:
+            d['summary_method'] = self.summary
+        if self.bold is not None:
+            d['bold'] = self.bold
+        if self.background is not None:
+            d['background'] = self.background
+        if (max_diff := self._max_diff()) is not None:
+            d['max_diff'] = max_diff
+        return d
```
[the remark this answers is not in the document]
Comment (takluyver) [reply damnit/ctxsupport/damnit_ctx.py:145]: These are already conceptually attributes for the summary, the thing we display in the table cell. We can also of course add attributes on the larger data behind that, stored in the HDF5 file, as well as the variable. Given that, namespacing them feels like overkill to me - I can't imagine that we'd want the same name for two different things. Does that seem reasonable? Or are you imagining a lot more attributes than me?

[the remark this answers is not in the document]
Comment (takluyver) [reply damnit/ctxsupport/damnit_ctx.py:145]: I could imagine supporting data attributes in `Cell` as well, but I'd be inclined to keep that fairly flat for clarity & checkability in the context file, even if it means a lot of very specific parameters.

File: damnit/backend/db.py [hunk 2]
```diff
@@ -16,11 +17,12 @@
 
 log = logging.getLogger(__name__)
 
-V1_SCHEMA = """
+V2_SCHEMA = """
```
Comment (JamesWrigley) [inline damnit/backend/db.py:20]: Let's add some brief docs here: https://damnit.readthedocs.io/en/latest/internals/#data-storage

File: damnit/ctxsupport/damnit_ctx.py [hunk 3]
```diff
@@ -69,3 +79,67 @@ def annotations(self):
...
+            if not re.match(r'#[0-9A-Fa-f]{6}', c):
+                raise ValueError("Colour string should be hex code (like '#ffcc00')")
+            b = bytes.fromhex(c[1:])
+            return np.frombuffer(b, dtype=np.uint8)
+        elif isinstance(c, Sequence):
+            if not len(c) == 3:
+                raise TypeError(f"Wrong number of values ({len(c)}) for R,G,B")
+            if not all(0 <= v <= 255 for v in c):
+                raise ValueError("Colour values must be 0 - 255")
+            return np.array(c, dtype=np.uint8)
```
Comment (JamesWrigley) [inline damnit/ctxsupport/damnit_ctx.py:116]: Is there a benefit to storing it as a uint8 array? I know that's an appropriate type, but I don't think there's a technical reason in this case, and if we stored it as a `list` instead we wouldn't need to handle ndarray's specially in `get_attrs()`.

File: damnit/backend/extract_data.py [hunk 4]
```diff
@@ -163,12 +163,26 @@ def get_dset_value(ds):
...
+    def get_attrs(ds):
+        d = {}
+        for name, value in ds.attrs.items():
+            if name in {"max_diff", "summary_method"}:
+                continue  # These are stored separately
+
+            if isinstance(value, np.ndarray):
+                value = value.tolist()
+            elif isinstance(value, np.generic):  # Scalar
+                value = value.item()
```
Comment (JamesWrigley) [inline damnit/backend/extract_data.py:175]: I think it'd be nice to move this logic into `Cell`.

File: damnit/gui/table.py [hunk 5]
```diff
@@ -387,7 +388,7 @@ def comment_item(self, text, comment_id=None):
             item.setData(comment_id, COMMENT_ID_ROLE)
         return item
 
-    def new_item(self, value, column_id, max_diff=0):
+    def new_item(self, value, column_id, max_diff=0, attrs=None):
```
Comment (JamesWrigley) [inline damnit/gui/table.py:391]: Nitpicking, does it even make sense for these arguments to have default values? I think we always want to pass them.

File: tests/test_backend.py [hunk 6]
```diff
@@ -429,6 +430,31 @@ def bad(run):
         assert set(f) == {".reduced", "good", "start_time"}
         assert set(f[".reduced"]) == {"good", "start_time"}
 
+def test_results_cell(mock_run, tmp_path):
+    # Test returning an object we can't save in HDF5
```
Comment (JamesWrigley) [inline tests/test_backend.py:434]: Hmmmm :stuck_out_tongue:

File: tests/test_backend.py [hunk 7]
```diff
@@ -429,6 +430,31 @@ def bad(run):
...
+    bad_obj_ctx = mkcontext(ctx_code)
+    results = bad_obj_ctx.execute(mock_run, 1000, 123, {})
+    results_hdf5_path = tmp_path / 'results.h5'
+    results.save_hdf5(results_hdf5_path)
+    with h5py.File(results_hdf5_path) as f:
+        assert f['.reduced/var2'][()] == 45
+        assert f['.reduced/var2'].attrs['bold'] == False
+        np.testing.assert_array_equal(
+            f['.reduced/var2'].attrs['background'], [0xbb, 0xdd, 0xdd]
+        )
```
Comment (JamesWrigley) [inline tests/test_backend.py:456]: It'd be good to have a test for a custom `summary_value` too.

**Review by JamesWrigley — CHANGES_REQUESTED** [review]

File: damnit/ctxsupport/damnit_ctx.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JamesWrigley) [reply damnit/ctxsupport/damnit_ctx.py:145]: Whatever we decide to make the user-facing API, we can always save them differently. Though I do lean towards sticking with a flat structure.

File: damnit/ctxsupport/damnit_ctx.py (hunk 3 above)
Comment (takluyver) [reply damnit/ctxsupport/damnit_ctx.py:116]: Between this code (in the context process) and `get_attrs` it's stored as an HDF5 attribute, which h5py will always read as a numpy array.

File: damnit/backend/extract_data.py (hunk 4 above)
Comment (takluyver) [reply damnit/backend/extract_data.py:175]: That doesn't work for much the same reason - the boundary between the context process (containing `Cell`) & this code is an HDF5 file, and h5py will return Numpy values.

I do want to revisit this layer as well, but I didn't want to do it as part of this PR.

File: damnit/gui/table.py (hunk 5 above)
Comment (takluyver) [reply damnit/gui/table.py:391]: You're right, we never call it without them. I think I changed this around a few times while I was working on that refactoring.

File: damnit/backend/db.py (hunk 2 above)
Comment (takluyver) [reply damnit/backend/db.py:20]: Good point, done.

File: tests/test_backend.py (hunk 6 above)
Comment (takluyver) [reply tests/test_backend.py]: Thanks, good catch.

File: tests/test_backend.py (hunk 7 above)
Comment (takluyver) [reply tests/test_backend.py:459]: It is good, adding it immediately caught a bug. :slightly_smiling_face:

**Comment by takluyver** [comment]
I've rebased to clear up merge conflicts with #262.

**Review by JamesWrigley — APPROVED** [review]
I think this will need a rebase, but otherwise LGTM!

**Pull request merged**